README.md
# pgslam
## introduction
pgslam (pose graph slam) is a slam package based on ROS. pgslam reads laser scans and odometry(tf) and publish occupancy map and pose(tf). pgslam dance icp algorithm to get pose as constrain and give these constrains to isam to construct a graph.
## installation
1. install ros indigo
 1. sudo sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
 2. sudo apt-key adv --keyserver hkp://ha.pool.sks-keyservers.net --recv-key 421C365BD9FF1F717815A3895523BAEEB01FA116
 3. sudo apt-get update
 4. sudo apt-get install ros-indigo-desktop-full
2. install isam
 1. wget http://people.csail.mit.edu/kaess/isam/isam_v1_7.tgz
 2. tar xzvf isam_V1_7.tgz
 3. cd isam_V1_7
 4. make
 5. sudo make install
3. create catkin workspace
 1. mkdir -p ~/catkin_ws/src
 2. cd ~/cattkin_ws/src/
 3. catkin_init_workspace
5. git clone https://github.com/yukunlinykl/pgslam ~/catkin_ws/src/pgslam
6. cd ~/catkin_ws && catkin_make
7. cd ~/catkin_ws && catkin_make install

## run
1. source /opt/ros/indigo/setup.bash
2. source catkin_ws/install/setup.bash
3. roslaunch pgslam playbag.launch


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package pgslam
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.9.10 (2016-12-28)
-------------------
* upload to github
* Contributors: Yu Kunlin


src/pgslam.cc
#include <pgslam/pgslam.h>
#include <pgslam/kdtree2d.h>

#include <iomanip>
#include <iostream>
#include <sstream>
#include <float.h>
#include <sys/time.h>

#include <Eigen/Eigen>

using namespace pgslam;

Pose2D::Pose2D ()
{
	this->x = 0;
	this->y = 0;
	this->theta = 0;
}

Pose2D::Pose2D (double x, double y, double theta)
{
	this->x = x;
	this->y = y;
	this->theta = theta;
	while (this->theta<-M_PI) this->theta += 2 * M_PI;
	while (this->theta> M_PI) this->theta -= 2 * M_PI;
}

void Pose2D::set_x(double x)
{
	this->x = x;
}

void Pose2D::set_y(double y)
{
	this->y = y;
}

void Pose2D::set_theta(double theta)
{
	this->theta = theta;
	while (this->theta<-M_PI) this->theta += 2 * M_PI;
	while (this->theta> M_PI) this->theta -= 2 * M_PI;
}

double Pose2D::get_x()
{
	return x;
}

double Pose2D::get_y()
{
	return y;
}

double Pose2D::get_theta ()
{
	return theta;
}

Pose2D Pose2D::operator + (Pose2D p)
{
	Eigen::Vector2d v1(x,y);
	Eigen::Vector2d v2(p.x,p.y);
	Eigen::Rotation2D<double> rot(theta);
	v2 = rot * v2;
	Eigen::Vector2d v = v1 + v2;
	return Pose2D (v.x(), v.y(), theta+p.theta);
}

Pose2D Pose2D::inverse ()
{
	Eigen::Vector2d v(-x,-y);
	Eigen::Rotation2D<double> rot(-theta);
	v = rot * v;
	return Pose2D (v.x(), v.y(), -theta);
}

Pose2D Pose2D::operator - (Pose2D p)
{
	return p.inverse() + *this;
}

Eigen::Vector2d Pose2D::pos ()
{
	return Eigen::Vector2d (x,y);
}

std::string Pose2D::to_string ()
{
	std::stringstream ss;
	ss << "x:";
	ss << std::setw (7) << setiosflags(std::ios::fixed) << std::setprecision (4);
	ss << x;
	ss << " y:";
	ss << std::setw (7) << setiosflags(std::ios::fixed) << std::setprecision (4);
	ss << y;
	ss << " theta:";
	ss << std::setw (7) << setiosflags(std::ios::fixed) << std::setprecision (4);
	ss << theta;
	return ss.str();
}

Echo::Echo (double range, double angle, double intensity, int64_t time_stamp)
{
	this->range = range;
	this->angle = angle;
	this->intensity = intensity;
	this->time_stamp = time_stamp;
}

double Echo::get_range ()
{
	return range;
}

double Echo::get_angle ()
{
	return angle;
}

double Echo::get_intensity ()
{
	return intensity;
}

int64_t Echo::get_time_stamp ()
{
	return time_stamp;
}

Eigen::Vector2d Echo::get_point ()
{
	double x = range * cos(angle);
	double y = range * sin(angle);
	return Eigen::Vector2d (x,y);
}

LaserScan::LaserScan (std::vector<Echo> echos)
{
	for (size_t i=0; i<echos.size(); i++)
		points_self.push_back (echos[i].get_point());
	world_transformed_flag = false;

	match_threshold = 0.1;
	dist_threshold = 1.0;
}

LaserScan::LaserScan (std::vector<Echo> echos, Pose2D pose)
{
	for (size_t i=0; i<echos.size(); i++)
		points_self.push_back (echos[i].get_point());
	this->pose = pose;
	world_transformed_flag = false;

	match_threshold = 0.1;
	dist_threshold = 1.0;
}

Pose2D LaserScan::get_pose ()
{
	return pose;
}

void LaserScan::set_pose (Pose2D pose)
{
	this->pose = pose;
	world_transformed_flag = false;
}

const std::vector<Eigen::Vector2d> & LaserScan::get_points ()
{
	UpdateToWorld ();
	return points_world;
}

void LaserScan::UpdateToWorld ()
{
	if (world_transformed_flag) return;

	std::vector<Eigen::Vector2d>().swap (points_world);

	max_x = 0.0;
	min_x = 0.0;
	max_y = 0.0;
	min_y = 0.0;

	Eigen::Rotation2D<double> rot (pose.theta);
	Eigen::Vector2d move = pose.pos();
	for (size_t i=0; i<points_self.size(); i++) {
		Eigen::Vector2d p = rot * points_self[i] + move;
		points_world.push_back (p);
		if (p.x()>max_x) max_x = p.x();
		if (p.x()<min_x) min_x = p.x();
		if (p.y()>max_y) max_y = p.y();
		if (p.y()<min_y) min_y = p.y();
	}

	world_transformed_flag = true;
}

/*
int LaserScan::nearest (std::vector<Eigen::Vector2d> &v, Eigen::Vector2d p)
{
	double dist = DBL_MAX;
	int index_rough=-1,index;
	int speed_up_num = 10;
	for (size_t i=0; i<v.size(); i+=speed_up_num) {
		if ((v[i]-p).norm()<dist) {
			dist = (v[i]-p).norm();
			index_rough = i;
		}
	}
	index = index_rough;
	for (int j=index_rough-speed_up_num; j<index_rough+speed_up_num; j++) {
		int i = (j+v.size())%v.size();
		if ((v[i]-p).norm()<dist) {
			dist = (v[i]-p).norm();
			index = i;
		}
	}
	return index;
}
*/

std::vector<Eigen::Vector2d> LaserScan::transform (const std::vector<Eigen::Vector2d> &v, Pose2D pose)
{
	std::vector<Eigen::Vector2d> v2 (v.size());
	Eigen::Rotation2D<double> rot (pose.theta);
	Eigen::Vector2d move (pose.x, pose.y);
	for (size_t i=0; i<v.size(); i++) {
		v2[i] = rot * v[i];
		v2[i] += move;
	}
	return v2;
}

Pose2D LaserScan::icp(std::vector<Eigen::Vector2d> scan_ref, std::vector<Eigen::Vector2d> scan, double *ratio, Pose2D reference_pose)
{
	if (scan_ref.size()<2 || scan.size()<2) {
		std::cout << "Error: scan.size() < 2";
		return reference_pose;
	}

	size_t insert_num = 7;
	auto scan_cache = scan_ref;
	scan_ref.resize(scan_cache.size()*insert_num);
	for (size_t i=0; i<scan_cache.size()-1; i++) {
		for (size_t j=0; j<insert_num; j++) {
			scan_ref[insert_num*i+j] = (scan_cache[i+1] - scan_cache[i]) / insert_num * j + scan_cache[i];
		}
	}

	kd_tree_2d::KDTree2D tree;
	tree.Construct (scan_ref);
	std::vector<Eigen::Vector2d> scan_origin = scan;
	Pose2D pose = reference_pose;
	for (int i=0; i<20; i++) {
		scan = transform (scan_origin, pose);

		std::vector<Eigen::Vector2d>    near = scan;                 // store the closest point
		std::vector< std::vector<int> >	trace_back(scan_ref.size()); // to trace some points have a same closest point
		std::vector<bool>		mask(scan.size());           // true: effective point; false: non-effective point;

		// search and save nearest point
		int match_count = 0;
		for (size_t i=0; i<scan.size(); i++) {
			Eigen::Vector2d point = scan[i];

			//int index = nearest (scan_ref,scan[i]);
			size_t index = tree.NearestIndex (scan[i]);

			if (index==-1) {
				if (ratio!=nullptr)
					*ratio = 0.0;
				return Pose2D();
			}
			trace_back[index].push_back(i);
			Eigen::Vector2d closest = scan_ref[index];

			double distance = (point-closest).norm();
			if (distance < match_threshold)
				match_count ++;
			if (distance < dist_threshold) {
				near[i] = closest;
				mask[i] = true;
			} else {
				mask[i] = false;
			}
		}
		if (ratio!=nullptr)
			*ratio = (double)match_count/scan.size();

		// disable the points have one same nearest point
		for (size_t i=0; i<trace_back.size(); i++)
			if (trace_back[i].size()>3)
				for (size_t j=0; j<trace_back[i].size(); j++) {
					mask[trace_back[i][j]] = false;
					near [trace_back[i][j]] = scan[trace_back[i][j]];
				}

		// disable the farest 10% point
		std::vector<double> max_distance;
		std::vector<int>    max_index;
		max_distance.resize (scan.size()/10,0.0);
		max_index.resize (scan.size()/10,0);
		for (size_t i=0; i<scan.size(); i++) {
			double distance = (scan[i]-near[i]).norm();
			for (size_t j=1; j<max_distance.size(); j++) {
				if (distance>max_distance[j]) {
					max_distance[j-1] = max_distance[j];
					max_index[j-1] = max_index[j];
					if (j==max_distance.size()-1) {
						max_distance[j] = distance;
						max_index[j] = i;
					}
				}
				else {
					max_distance[j-1] = distance;
					max_index[j-1] = i;
					break;
				}
			}
		}
		for (size_t i=1; i<max_index.size(); i++)
			mask[max_index[i]] = false;

		// calc center
		Eigen::Vector2d center(0,0);
		int count = 0;
		for (size_t i=0; i<scan.size(); i++)
			if (mask[i]) {
				center += scan[i];
				count ++;
			}
		if (count==0) {
			std::cout << "Error: no valid point, return reference pose." << std::endl;
			if (ratio!=nullptr)
				*ratio = 0.0;
			return reference_pose;
		}
		center /= count;

		// calc the tramsform
		Eigen::Vector2d move = Eigen::Vector2d(0.0,0.0);
		double rot = 0.0;
		for (size_t i=0; i<scan.size(); i++) {
			if (mask[i]==false) continue;
			Eigen::Vector2d delta = near[i] - scan[i];
			double length = delta.norm();
			if (length>0) {
				delta.normalize();
				delta *= (length<0.05)?length:sqrt(length*20)/20;
			}
			move += delta;
			Eigen::Vector2d p = scan[i] - center;
			Eigen::Vector2d q = near[i] - center;
			if (p.norm()<DBL_EPSILON*2) continue;

			rot += (p.x()*q.y() - p.y()*q.x()) / p.norm() / sqrt(p.norm());
		}
		move /= count;
		rot /= count;

		// speed up
		move *= 2.0;
		rot *= 1.0;

		pose.x += move.x();
		pose.y += move.y();
		pose.theta += rot;
	}
	return pose;
}

Pose2D LaserScan::ICP (const LaserScan &scan_, double *ratio)
{
	LaserScan scan = scan_;
	return icp (points_self, scan.points_self, ratio, scan.pose-pose);
}

double LaserScan::get_max_x_in_world ()
{
	UpdateToWorld ();
	return max_x;
}

double LaserScan::get_min_x_in_world ()
{
	UpdateToWorld ();
	return min_x;
}

double LaserScan::get_max_y_in_world ()
{
	UpdateToWorld ();
	return max_y;
}

double LaserScan::get_min_y_in_world ()
{
	UpdateToWorld ();
	return min_y;
}

#ifdef USE_ISAM
GraphSlam::GraphSlam ()
{
	slam = new isam::Slam();
}

bool GraphSlam::check (size_t id)
{
	// size enough
	if (id<pose_nodes.size()) {
		if (pose_nodes[id]==NULL) {	// already been removed
			pose_nodes[id] = new isam::Pose2d_Node();
			return true;
		}
		else return false;	// still there
  }

	// create new space
	int size = pose_nodes.size();
	pose_nodes.resize(id+1);
	for (size_t i=size; i<id+1; i++)
		pose_nodes[i] = new isam::Pose2d_Node();
	return true;
}

void GraphSlam::remove (size_t node_id)
{
	slam->remove_node (pose_nodes[node_id]);
	delete pose_nodes[node_id];
	pose_nodes[node_id] = NULL;
	slam->batch_optimization();
}

void GraphSlam::clear ()
{
	delete slam;
	slam = new isam::Slam();
	std::vector<isam::Pose2d_Node*>().swap(pose_nodes);
}

std::vector< std::pair<size_t, Pose2D> > GraphSlam::get_nodes ()
{
	std::vector< std::pair<size_t, Pose2D> > pose_id;
	// store in response
	for (size_t i=0; i<pose_nodes.size(); i++) {
		if (pose_nodes[i]==NULL) continue;
		Pose2D pose;
		pose.set_x(pose_nodes[i]->value().x());
		pose.set_y(pose_nodes[i]->value().y());
		pose.set_theta(pose_nodes[i]->value().t());
		pose_id.push_back (std::pair<size_t, Pose2D>(i,pose));
	}
	return pose_id;
}

std::vector< std::pair<Eigen::Vector2d, Eigen::Vector2d> > GraphSlam::get_factors ()
{
	std::vector< std::pair<Eigen::Vector2d, Eigen::Vector2d> > factors;
	const std::list<isam::Factor*> factors_ = slam->get_factors();
	for (std::list<isam::Factor*>::const_iterator it=factors_.begin(); it!=factors_.end(); it++) {
		std::vector<isam::Node*> nodes = (*it)->nodes();
		if (nodes.size()==1) continue;
		Eigen::Vector2d first;
		first.x() = ((isam::Pose2d_Node*)nodes[0])->value().x();
		first.y() = ((isam::Pose2d_Node*)nodes[0])->value().y();
		Eigen::Vector2d second;
		second.x() = ((isam::Pose2d_Node*)nodes[1])->value().x();
		second.y() = ((isam::Pose2d_Node*)nodes[1])->value().y();
		factors.push_back (std::pair<Eigen::Vector2d,Eigen::Vector2d>(first,second));
	}
	return factors;
}

void GraphSlam::AddPose2dFactor (size_t node_id, Pose2D pose_ros, double cov)
{
	isam::Pose2d pose (pose_ros.get_x(), pose_ros.get_y(), pose_ros.get_theta());
	if (cov<=0) {
		cov = 1.0;
	}

	// check new node
	bool ret = check (node_id);
	if (ret) slam->add_node (pose_nodes[node_id]);

	// add factor
	isam::Noise noise = isam::Information (cov * isam::eye(3));
	isam::Pose2d_Factor * factor = new isam::Pose2d_Factor(pose_nodes[node_id], pose, noise);
	slam->add_factor(factor);

	//slam->batch_optimization();
}

void GraphSlam::AddPose2dPose2dFactor (size_t node_id_ref, size_t node_id, Pose2D pose_ros, double cov)
{
	isam::Pose2d pose (pose_ros.get_x(), pose_ros.get_y(), pose_ros.get_theta());
	if (cov<=0) {
		cov = 1.0;
	}

	// check new node
	bool ret = check (node_id_ref);
	if (ret) slam->add_node (pose_nodes[node_id_ref]);
	ret = check (node_id);
	if (ret) slam->add_node (pose_nodes[node_id]);

	// add factor
	isam::Noise noise = isam::Information (cov * isam::eye(3));
	isam::Pose2d_Pose2d_Factor * factor = new isam::Pose2d_Pose2d_Factor(pose_nodes[node_id_ref], pose_nodes[node_id], pose, noise);
	slam->add_factor(factor);

	//slam->batch_optimization();
}

void GraphSlam::Optimization ()
{
	slam->batch_optimization();
}
#endif

Slam::Slam ()
{
	keyscan_threshold = 0.4;
	factor_threshold = 0.9;
	pose_update_callback = nullptr;
	map_update_callback = nullptr;
}

void Slam::set_keyscan_threshold (double keyscan_threshold)
{
	this->keyscan_threshold = keyscan_threshold;
	if (keyscan_threshold*2 > factor_threshold)
		factor_threshold = keyscan_threshold*2;
}

void Slam::set_factor_threshold (double factor_threshold)
{
	this->factor_threshold = factor_threshold;
	if (keyscan_threshold*2 > factor_threshold)
		keyscan_threshold = factor_threshold/2;
}

Pose2D Slam::get_pose ()
{
	return pose;
}

const std::vector<LaserScan> & Slam::get_scans ()
{
	return scans;
}

#ifdef USE_ISAM
std::vector< std::pair<Eigen::Vector2d, Eigen::Vector2d> > Slam::get_factors ()
{
	return graph_slam.get_factors();
}
#endif

Pose2D Slam::EncoderToPose2D (double left, double right, double tread)
{
	double theta = (right-left) / tread;
	double theta_2 = theta / 2.0;
	double arc = (right+left) / 2.0;
	double radius = arc / theta;
	double secant = 2 * sin (theta_2) * radius;
	if (theta == 0)
		secant = arc;
	double x = secant * cos (theta_2);
	double y = secant * sin (theta_2);

	return Pose2D (x,y,theta);
}

void Slam::UpdatePoseWithPose (Pose2D pose)
{
	this->pose = this->pose + pose;
}

void Slam::UpdatePoseWithEncoder (double left, double right, double tread)
{
	pose = pose + EncoderToPose2D (left,right,tread);
	if (pose_update_callback != nullptr)
		pose_update_callback ();
}

void Slam::UpdatePoseWithLaserScan (const LaserScan &scan_)
{
	LaserScan scan = scan_;
	scan.set_pose (pose);

	// first scan
	if (scans.empty()) {
		scans.push_back (scan);
#ifdef USE_ISAM
		graph_slam.AddPose2dFactor (0,pose,1);
#endif
		std::cout << "add key scan " << scans.size() << ": " << pose.to_string() << std::endl;
		if (map_update_callback != nullptr)
			map_update_callback ();
		return;
	}

	// search for the closest scan
	LaserScan *closest_scan = &(scans[0]);
	double min_dist = DBL_MAX;
	for (size_t i=0; i<scans.size(); i++) {
		double dist = (scans[i].get_pose().pos() - scan.get_pose().pos()).norm();
		double delta_theta = fabs(scans[i].get_pose().get_theta() - scan.get_pose().get_theta());;
		while (delta_theta<-M_PI) delta_theta += 2 * M_PI;
		while (delta_theta> M_PI) delta_theta -= 2 * M_PI;
		delta_theta *= keyscan_threshold / (M_PI_4 * 3.0);

		dist = sqrt(dist*dist + delta_theta*delta_theta);
		if (dist < min_dist) {
			min_dist = dist;
			closest_scan = &(scans[i]);
		}
	}


	if (min_dist < keyscan_threshold) {
		// update pose
		double ratio;
		Pose2D pose_delta = closest_scan->ICP (scan, &ratio);
		pose = closest_scan->get_pose() + pose_delta;
	}
	else {
		// add key scan
#ifdef USE_ISAM
		for (size_t i=0; i<scans.size(); i++) {
			double distance = (pose.pos() - scans[i].get_pose().pos()).norm();
			if (distance<factor_threshold) {
				double ratio;
				Pose2D pose_delta = scans[i].ICP (scan, &ratio);
				graph_slam.AddPose2dPose2dFactor (i, scans.size(), pose_delta, ratio);
				if (pose_update_callback != nullptr)
					pose_update_callback ();
			}
		}
		graph_slam.Optimization ();

		auto nodes = graph_slam.get_nodes ();
		for (size_t i=0; i<nodes.size(); i++) {
			// update pose of node
			for (size_t j=0; j<scans.size(); j++) {
				if (j != nodes[i].first) continue;
				scans[j].set_pose (nodes[i].second);
				break;
			}
			// add the new scan with pose
			if (nodes[i].first == scans.size()) {
				pose = nodes[i].second;
				scan.set_pose (nodes[i].second);
				scans.push_back (scan);
			}

		}
#else
		scans.push_back (scan);
#endif
		std::cout << "add key scan " << scans.size() << ": " << pose.to_string() << std::endl;

		if (map_update_callback != nullptr)
			map_update_callback ();
	}
	if (pose_update_callback != nullptr)
		pose_update_callback ();
}

void Slam::RegisterPoseUpdateCallback (void(*f)(void))
{
	pose_update_callback = f;
}
void Slam::RegisterMapUpdateCallback (void(*f)(void))
{
	map_update_callback = f;
}



src/kdtree2d.cc
#include <pgslam/kdtree2d.h>
#include <iostream>

using namespace kd_tree_2d;

Node::Node (Eigen::Vector2d point, size_t index, int dir)
{
	this->point = point;
	this->index = index;
	left = nullptr;
	right = nullptr;
	this->dir = dir;
}

Eigen::Vector2d Node::get_point ()
{
	return point;
}

size_t Node::get_index ()
{
	return index;
}

void Node::insert (Eigen::Vector2d point, size_t index)
{
	if (point(dir) <= this->point(dir)) {
		if (left==nullptr) {
			left = new Node (point, index, (dir+1)%2);
		} else {
			left->insert (point, index);
		}
	}
	else {
		if (right==nullptr) {
			right = new Node (point, index, (dir+1)%2);
		} else {
			right->insert (point, index);
		}
	}
}

Node * Node::Nearest (Eigen::Vector2d point)
{
	Node *near_side, *far_side;
	Node *nearside_best=nullptr;
	Node *farside_best=nullptr;
	if (left==nullptr && right==nullptr)
		return this;
	if (point(dir) <= this->point(dir)) {
		near_side = left;
		far_side = right;
	} else {
		near_side = right;
		far_side = left;
	}

	if (near_side!=nullptr) {
		nearside_best = near_side->Nearest (point);
	}

	if ( nearside_best==nullptr || (nearside_best->point - point).norm() > fabs(point(dir)-this->point(dir))) {
		if (far_side!=nullptr) {
			farside_best = far_side->Nearest (point);
		}
	}
	Node * best = this;
	if (nearside_best!=nullptr && (nearside_best->point-point).norm() < (best->point-point).norm()) {
		best = nearside_best;
	}
	if (farside_best!=nullptr && (farside_best->point-point).norm() < (best->point-point).norm()) {
		best = farside_best;
	}
	return best;
}

Node::~Node ()
{
	if (left!=nullptr)
		delete left;
	if (right!=nullptr)
		delete right;
}

KDTree2D::KDTree2D ()
{
	root = nullptr;
}

void KDTree2D::insert (Eigen::Vector2d point, size_t index)
{
	if (root==nullptr) {
		root = new Node (point,index,0);
	} else {
		root->insert (point,index);
	}
}

void KDTree2D::Construct (const std::vector<Eigen::Vector2d> &points)
{
	std::vector<size_t> index;
	for (size_t i=0; i<points.size(); i++)
		index.push_back (i);
	srand (time(NULL));
	for (size_t i=0; i<index.size()/2; i++) {
		size_t a = rand()%index.size();
		size_t b = rand()%index.size();
		size_t t = index[a];
		index[a] = index[b];
		index[b] = t;
	}
	for (size_t i=0; i<index.size(); i++) {
		insert (points[index[i]],index[i]);
	}
}

Eigen::Vector2d KDTree2D::Nearest (Eigen::Vector2d point)
{
	assert (root!=nullptr);
	auto result = root->Nearest (point);
	if (result==nullptr) {
		std::cout << "error" << std::endl;
	}
	else {
		return result->get_point();
	}
}

size_t KDTree2D::NearestIndex (Eigen::Vector2d point)
{
	assert (root!=nullptr);
	auto result = root->Nearest (point);
	if (result==nullptr) {
		std::cout << "error" << std::endl;
	}
	else {
		return result->get_index();
	}
}

KDTree2D::~KDTree2D ()
{
	delete root;
}


src/pgslam_node.cc
#include <string>

#include <Eigen/Eigen>

#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <nav_msgs/OccupancyGrid.h>
#include <tf/transform_listener.h>
#include <tf/transform_broadcaster.h>
#include <visualization_msgs/Marker.h>

#include <pgslam/pgslam.h>

using namespace std;
using namespace ros;

Publisher node_pub;
Publisher factor_pub;
Publisher map_pub;
nav_msgs::OccupancyGrid global_map;
tf::TransformListener * plistener;

pgslam::Slam slam;

string map_frame  = "map";
string odom_frame = "odom";
string base_frame = "base_link";
double keyscan_threshold = 0.4;
double factor_threshold = 0.9;

void draw_graph ()
{
	visualization_msgs::Marker points;
	points.header.frame_id = map_frame;
	points.header.stamp = Time::now();
	points.ns = "points_and_lines";
	points.action = visualization_msgs::Marker::ADD;
	points.pose.orientation.w = 1.0;
	points.id = 0;
	points.type = visualization_msgs::Marker::POINTS;
	points.scale.x = 0.05;
	points.scale.y = 0.05;
	points.color.g = 0.5f;
	points.color.a = 1.0;

	auto scans = slam.get_scans();
	scans[0].set_pose(pgslam::Pose2D());
	for (int i=0; i<scans.size(); i++) {
		geometry_msgs::Point p;
		p.x = scans[i].get_pose().pos().x();
		p.y = scans[i].get_pose().pos().y();
		p.z = 0;
		points.points.push_back(p);
	}
	node_pub.publish(points);

	visualization_msgs::Marker line_list;
	line_list.header.frame_id = map_frame;
	line_list.header.stamp = Time::now();
	line_list.ns = "points_and_lines";
	line_list.action = visualization_msgs::Marker::ADD;
	line_list.pose.orientation.w = 1.0;
	line_list.id = 1;
	line_list.type = visualization_msgs::Marker::LINE_LIST;
	line_list.scale.x = 0.01;
	line_list.color.r = 0.5;
	line_list.color.a = 1.0;

#ifdef USE_ISAM
	vector< pair<Eigen::Vector2d, Eigen::Vector2d> > factors = slam.get_factors ();
	for (size_t i=0; i<factors.size(); i++) {
		geometry_msgs::Point first;
		first.x = factors[i].first.x();
		first.y = factors[i].first.y();
		line_list.points.push_back (first);
		geometry_msgs::Point second;
		second.x = factors[i].second.x();
		second.y = factors[i].second.y();
		line_list.points.push_back (second);
	}
	factor_pub.publish(line_list);
#endif
}

void draw_map ()
{
	double resolution = 0.05;
	double draw_range = 6.0;
	param::get ("~resolution",resolution);
	param::get ("~draw_range",draw_range);

	double max_x = 0.0;
	double min_x = 0.0;
	double max_y = 0.0;
	double min_y = 0.0;
	auto scans = slam.get_scans ();
	for (size_t i=0; i<scans.size(); i++) {
		if (max_x < scans[i].get_max_x_in_world()) max_x = scans[i].get_max_x_in_world();
		if (min_x > scans[i].get_min_x_in_world()) min_x = scans[i].get_min_x_in_world();
		if (max_y < scans[i].get_max_y_in_world()) max_y = scans[i].get_max_y_in_world();
		if (min_y > scans[i].get_min_y_in_world()) min_y = scans[i].get_min_y_in_world();
	}
	double max_x_i = (int)((max_x+1.0)*10)/10.0;
	double min_x_i = (int)((min_x-1.0)*10)/10.0;
	double max_y_i = (int)((max_y+1.0)*10)/10.0;
	double min_y_i = (int)((min_y-1.0)*10)/10.0;
	int width = (max_x_i-min_x_i)/resolution;
	int height= (max_y_i-min_y_i)/resolution;

	// new an eigen map
	Eigen::MatrixXi emap (width, height);
	for (int i=0; i<width; i++)
		for (int j=0; j<height; j++)
			emap(i,j) = -1;

	// draw map into eigen map
	Eigen::Vector2d source (min_x_i, min_y_i);
	for (int i=0; i<scans.size(); i++) {	// for every scan
		Eigen::Vector2d origin = scans[i].get_pose().pos();
		auto points = scans[i].get_points();
		for (int j=1; j<points.size(); j++) {	// for every point
			Eigen::Vector2d v = points[j] - origin;
			if (v.norm()>draw_range)
				v = v.normalized() * (draw_range + resolution);
			int steps = v.norm()/resolution;
			Eigen::Vector2d step = v/steps/resolution;
			Eigen::Vector2d current = (origin -source)/resolution;
			for (int i=0; i<steps; i++) {	// for every step
				if (emap(current.x(),current.y())==-1)
					emap(current.x(),current.y()) = 30;// white
				else	emap(current.x(),current.y()) = (int)(emap(current.x(),current.y())*0.8);
				current += step;
			}
			if (v.norm()<draw_range)
				if (emap(current.x(),current.y())==-1)
					emap(current.x(),current.y()) = 100;	// black
				else {
					emap(current.x(),current.y()) = (int)(emap(current.x(),current.y())*0.2);
					emap(current.x(),current.y()) += 100*0.8;
				}
		}
	}

	// copy eigen map to ros map
	global_map.header.stamp = Time::now();
	global_map.header.frame_id = map_frame;
	global_map.info.resolution = resolution;
	global_map.info.map_load_time = Time::now();

	global_map.info.width = width;
	global_map.info.height= height;

	global_map.info.origin.position.x = min_x_i;
	global_map.info.origin.position.y = min_y_i;
	global_map.info.origin.position.z = 0;
	global_map.info.origin.orientation.w = 1.0;

	global_map.data.resize(width*height);

	for (int i=0; i<width; i++)
		for (int j=0; j<height; j++)
			global_map.data[j*width+i] = emap(i,j);

	// publish
	map_pub.publish (global_map);
}

pgslam::Pose2D ListenPose2D (string target_frame, string source_frame)
{
	pgslam::Pose2D pose;
	// listen
	tf::StampedTransform transform;
	Time past = Time::now() - Duration(0.1);
	std::string s;
	try {
		plistener->waitForTransform(target_frame, source_frame, Time(0), Duration(2.0), Duration(0.01), &s);
		plistener->lookupTransform (target_frame, source_frame, Time(0), transform);
	} catch (tf::TransformException ex) {
		ROS_ERROR ("slam lookupTransform error: %s %s", ex.what(), s.c_str());
		return pose;
	}

	// calc pose
	pose.set_x(transform.getOrigin().x());
	pose.set_y(transform.getOrigin().y());

	double roll,pitch,yaw;
	transform.getBasis().getRPY(roll, pitch, yaw);

	pose.set_theta(yaw);

	return pose;
}

void BroadcastMapAndGraph ()
{
	draw_graph ();
	draw_map ();
}

void BroadcastPose ()
{
	pgslam::Pose2D map_pose = slam.get_pose();
	pgslam::Pose2D odom_pose = ListenPose2D (odom_frame, base_frame);
	pgslam::Pose2D delta = map_pose + odom_pose.inverse();

	tf::StampedTransform transform;
	transform.setOrigin( tf::Vector3(delta.pos().x(), delta.pos().y(), 0.0) );
	tf::Quaternion q;
	q.setRPY(0, 0, delta.get_theta());
	transform.setRotation(q);

	static tf::TransformBroadcaster br;
	br.sendTransform(tf::StampedTransform(transform, ros::Time::now(), map_frame, odom_frame));
}

pgslam::LaserScan RosLaserScan_T_PGSlamLaserScan (const sensor_msgs::LaserScan& msg)
{
	std::vector<pgslam::Echo> echos;
	size_t i=0;
	for (double angle=msg.angle_min; angle<=msg.angle_max; angle+=msg.angle_increment, i++) {
		echos.push_back (pgslam::Echo(msg.ranges[i],angle,msg.intensities[i],0));
	}
	return pgslam::LaserScan (echos);
}

void scanCallback (const sensor_msgs::LaserScan& msg)
{
	static pgslam::Pose2D odom_old;
	pgslam::Pose2D odom_new = ListenPose2D (odom_frame, base_frame);
	pgslam::Pose2D odom_delta = odom_new - odom_old;
	odom_old = odom_new;
	slam.UpdatePoseWithPose (odom_delta);
	slam.UpdatePoseWithLaserScan (RosLaserScan_T_PGSlamLaserScan(msg));
}

int main(int argc, char **argv)
{
	init (argc, argv, "pgslam");
	NodeHandle node;

	Subscriber scan_sub = node.subscribe ("/scan", 10, scanCallback);

	node_pub   = node.advertise<visualization_msgs::Marker> ("graph_node", 1);
	factor_pub = node.advertise<visualization_msgs::Marker> ("graph_factor", 1);
	map_pub    = node.advertise<nav_msgs::OccupancyGrid> ("/map",1);

	slam.RegisterMapUpdateCallback (BroadcastMapAndGraph);
	slam.RegisterPoseUpdateCallback (BroadcastPose);

	plistener = new tf::TransformListener();

	param::get ("~map_frame", map_frame );
	param::get ("~odom_frame",odom_frame);
	param::get ("~base_frame",base_frame);
	param::get ("~keyscan_threshold", keyscan_threshold);
	param::get ("~factor_threshold", factor_threshold);

	slam.set_keyscan_threshold (keyscan_threshold);
	slam.set_factor_threshold (factor_threshold);

	spin ();

	return 0;
}

